File: graph/FloydWarshall/java/FloydWarshall.java
import java.util.List;

public final class FloydWarshall {
    private FloydWarshall() {}

    public static final class Edge {
        public final int source;
        public final int target;
        public final long weight;

        public Edge(int source, int target, long weight) {
            this.source = source;
            this.target = target;
            this.weight = weight;
        }
    }

    public enum Status {
        OK,
        NEGATIVE_CYCLE,
        OVERFLOW
    }

    public static final class Result {
        public final Status status;
        public final Long[][] distances;

        public Result(Status status, Long[][] distances) {
            this.status = status;
            this.distances = distances;
        }
    }

    private static Long checkedAdd(long left, long right) {
        if (right > 0 && left > Long.MAX_VALUE - right) {
            return null;
        }
        if (right < 0 && left < Long.MIN_VALUE - right) {
            return null;
        }
        return left + right;
    }

    public static Result floydWarshall(int nodeCount, List<Edge> edges) {
        Long[][] distances = new Long[nodeCount][nodeCount];

        for (int node = 0; node < nodeCount; ++node) {
            distances[node][node] = 0L;
        }

        for (Edge edge : edges) {
            Long current = distances[edge.source][edge.target];
            if (current == null || edge.weight < current) {
                distances[edge.source][edge.target] = edge.weight;
            }
        }

        for (int intermediate = 0; intermediate < nodeCount; ++intermediate) {
            for (int source = 0; source < nodeCount; ++source) {
                Long left = distances[source][intermediate];
                if (left == null) {
                    continue;
                }

                for (int target = 0; target < nodeCount; ++target) {
                    Long right = distances[intermediate][target];
                    if (right == null) {
                        continue;
                    }

                    Long candidate = checkedAdd(left, right);
                    if (candidate == null) {
                        return new Result(Status.OVERFLOW, distances);
                    }

                    Long current = distances[source][target];
                    if (current == null || candidate < current) {
                        distances[source][target] = candidate;
                    }
                }
            }
        }

        for (int node = 0; node < nodeCount; ++node) {
            Long diagonal = distances[node][node];
            if (diagonal != null && diagonal < 0) {
                return new Result(Status.NEGATIVE_CYCLE, distances);
            }
        }

        return new Result(Status.OK, distances);
    }
}

File: graph/FloydWarshall/rust/src/floyd_warshall.rs
#[derive(Clone, Copy)]
pub struct Edge {
    pub source: usize,
    pub target: usize,
    pub weight: i64,
}

#[derive(Clone, Copy, PartialEq, Eq)]
pub enum Status {
    Ok,
    NegativeCycle,
    Overflow,
}

pub struct Result {
    pub status: Status,
    pub distances: Vec<Vec<Option<i64>>>,
}

pub fn floyd_warshall(node_count: usize, edges: &[Edge]) -> Result {
    let mut distances = vec![vec![None; node_count]; node_count];

    for (node, row) in distances.iter_mut().enumerate().take(node_count) {
        row[node] = Some(0_i64);
    }

    for edge in edges {
        match distances[edge.source][edge.target] {
            Some(current) if current <= edge.weight => {}
            _ => distances[edge.source][edge.target] = Some(edge.weight),
        }
    }

    for intermediate in 0..node_count {
        for source in 0..node_count {
            let Some(left) = distances[source][intermediate] else {
                continue;
            };

            for target in 0..node_count {
                let Some(right) = distances[intermediate][target] else {
                    continue;
                };

                let Some(candidate) = left.checked_add(right) else {
                    return Result {
                        status: Status::Overflow,
                        distances,
                    };
                };

                match distances[source][target] {
                    Some(current) if current <= candidate => {}
                    _ => distances[source][target] = Some(candidate),
                }
            }
        }
    }

    for (node, row) in distances.iter().enumerate().take(node_count) {
        if let Some(diagonal) = row[node] {
            if diagonal < 0 {
                return Result {
                    status: Status::NegativeCycle,
                    distances,
                };
            }
        }
    }

    Result {
        status: Status::Ok,
        distances,
    }
}
